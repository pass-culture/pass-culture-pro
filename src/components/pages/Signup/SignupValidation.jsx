import React, { PureComponent } from 'react'
import { Redirect, withRouter } from 'react-router-dom'
import { showNotification } from 'pass-culture-shared'
import { compose } from 'redux'
import { connect } from 'react-redux'
import { requestData } from 'redux-saga-data'

class SignupValidation extends PureComponent {
  componentDidMount() {
    const {
      dispatch,
      match: {
        params: { token },
      },
    } = this.props

    dispatch(
      requestData({
        apiPath: `/validate/user/${token}`,
        method: 'PATCH',
        handleSuccess: () => {
          dispatch(
            showNotification({
              text:
                'Votre compte a été créé. Vous pouvez vous connecter avec les identifiants que vous avez choisis.',
              type: 'success',
            })
          )
        },
        handleFail: (state, action) => {
          const {
            payload: { errors },
          } = action
          dispatch(
            showNotification({
              text: errors.global,
              type: 'danger',
            })
          )
        },
      })
    )
  }

  render() {
    return <Redirect to="/connexion" />
  }
}

export default SignupValidation
