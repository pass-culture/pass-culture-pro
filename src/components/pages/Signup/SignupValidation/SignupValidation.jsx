import PropTypes from 'prop-types'
import React, { PureComponent } from 'react'
import { Redirect } from 'react-router-dom'
import { requestData } from 'redux-saga-data'

import { redirectLoggedUser } from 'components/router/helpers'
import { campaignTracker } from 'tracking/mediaCampaignsTracking'

class SignupValidation extends PureComponent {
  constructor(props) {
    super(props)
    const { currentUser, history, isNewHomepageActive } = props
    redirectLoggedUser(history, currentUser, isNewHomepageActive)
  }

  componentDidMount() {
    campaignTracker.signUpValidation()

    const {
      dispatch,
      match: {
        params: { token },
      },
    } = this.props

    dispatch(this.buildRequestData(token))
  }

  buildRequestData = token => {
    return requestData({
      apiPath: `/validate/user/${token}`,
      method: 'PATCH',
      handleSuccess: this.notifySuccess(),
      handleFail: this.notifyFailure(),
    })
  }

  notifyFailure = () => {
    return (state, action) => {
      const {
        payload: { errors },
      } = action

      const { notifyError } = this.props
      notifyError(errors.global)
    }
  }

  notifySuccess = () => {
    return () => {
      const { notifySuccess } = this.props
      notifySuccess(
        'Votre compte a été créé. Vous pouvez vous connecter avec les identifiants que vous avez choisis.'
      )
    }
  }

  render() {
    return <Redirect to="/connexion" />
  }
}

SignupValidation.defaultProps = {
  currentUser: null,
}

SignupValidation.propTypes = {
  currentUser: PropTypes.shape(),
  dispatch: PropTypes.func.isRequired,
  history: PropTypes.func.isRequired,
  isNewHomepageActive: PropTypes.bool.isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      token: PropTypes.string.isRequired,
    }),
  }).isRequired,
  notifyError: PropTypes.func.isRequired,
  notifySuccess: PropTypes.func.isRequired,
}

export default SignupValidation
