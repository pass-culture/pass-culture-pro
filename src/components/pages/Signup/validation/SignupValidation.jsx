import React, { PureComponent } from 'react'
import { Redirect } from 'react-router-dom'
import { showNotification } from 'pass-culture-shared'
import { requestData } from 'redux-saga-data'

class SignupValidation extends PureComponent {
  buildRequestData = token => {
    return requestData({
      apiPath: `/validate/user/${token}`,
      method: 'PATCH',
      handleSuccess: this.notifySuccess(),
      handleFail: this.notifyFailure(),
    })
  }

  notifyFailure = () => {
    return (state, action) => {
      const {
        payload: { errors },
      } = action

      const { dispatch } = this.props
      dispatch(
        showNotification({
          text: errors.global,
          type: 'danger',
        })
      )
    }
  }

  notifySuccess = () => {
    return () => {
      const { dispatch } = this.props
      dispatch(
        showNotification({
          text:
            'Votre compte a été créé. Vous pouvez vous connecter avec les identifiants que vous avez choisis.',
          type: 'success',
        })
      )
    }
  }

  componentDidMount() {
    const {
      dispatch,
      match: {
        params: { token },
      },
    } = this.props

    dispatch(this.buildRequestData(token))
  }

  render() {
    return <Redirect to="/connexion" />
  }
}

export default SignupValidation
