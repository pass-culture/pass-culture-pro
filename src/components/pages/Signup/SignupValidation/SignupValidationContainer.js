import { compose } from 'redux'
import { connect } from 'react-redux'
import SignupValidation from './SignupValidation'
import withRouter from 'react-router-dom/es/withRouter'

export default compose(
  withRouter,
  connect()
)(SignupValidation)
