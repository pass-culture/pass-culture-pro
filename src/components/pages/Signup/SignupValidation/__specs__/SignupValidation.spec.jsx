import SignupValidation from '../SignupValidation'
import { mount, shallow } from 'enzyme'
import React from 'react'
import { Redirect, Router } from 'react-router-dom'
import { createBrowserHistory } from 'history'

jest.mock('redux-thunk-data', () => {
  const { requestData } = jest.requireActual('fetch-normalize-data')

  return {
    requestData,
  }
})

describe('src | components | pages | Signup | validation', () => {
  let history
  let dispatch
  let props

  beforeEach(() => {
    history = createBrowserHistory()
    dispatch = jest.fn()
    props = {
      dispatch,
      match: {
        params: {
          token: 'AAA',
        },
      },
    }
  })

  it('should match the snapshot', () => {
    // when
    const wrapper = shallow(<SignupValidation {...props} />)

    // then
    expect(wrapper).toMatchSnapshot()
  })

  it('should render a Redirect component', () => {
    // when
    const wrapper = mount(
      <Router history={history}>
        <SignupValidation {...props} />
      </Router>
    )

    // then
    const redirect = wrapper.find(Redirect)
    expect(redirect).toHaveLength(1)
    expect(redirect.prop('to')).toBe('/connexion')
  })

  it('should dispatch an action to verify validity of user token', () => {
    // when
    mount(
      <Router history={history}>
        <SignupValidation {...props} />
      </Router>
    )

    // then
    expect(dispatch.mock.calls[0][0]).toStrictEqual({
      config: {
        apiPath: '/validate/user/AAA',
        handleFail: expect.any(Function),
        handleSuccess: expect.any(Function),
        method: 'PATCH',
      },
      type: 'REQUEST_DATA_PATCH_/VALIDATE/USER/AAA',
    })
  })

  describe('notifySuccess', () => {
    it('should display a notification with success message', () => {
      // given
      const wrapper = shallow(<SignupValidation {...props} />)
      const notifySuccess = wrapper.instance().notifySuccess()

      // when
      notifySuccess()

      // then
      const notifySuccessCall = dispatch.mock.calls[1][0]
      expect(notifySuccessCall).toStrictEqual({
        patch: {
          text:
            'Votre compte a été créé. Vous pouvez vous connecter avec les identifiants que vous avez choisis.',
          type: 'success',
        },
        type: 'SHOW_NOTIFICATION',
      })
    })
  })

  describe('notifyFailure', () => {
    it('should display a notification with error message', () => {
      // given
      const wrapper = shallow(<SignupValidation {...props} />)
      const notifyFailure = wrapper.instance().notifyFailure()
      const state = {}
      const action = {
        payload: {
          errors: {
            global: ['error1', 'error2'],
          },
        },
      }

      // when
      notifyFailure(state, action)

      // then
      const notifyFailureCall = dispatch.mock.calls[1][0]
      expect(notifyFailureCall).toStrictEqual({
        patch: {
          text: ['error1', 'error2'],
          type: 'danger',
        },
        type: 'SHOW_NOTIFICATION',
      })
    })
  })
})
