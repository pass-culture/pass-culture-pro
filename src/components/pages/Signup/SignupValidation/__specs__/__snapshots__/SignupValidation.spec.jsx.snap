// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src | components | pages | Signup | validation should match the snapshot 1`] = `ShallowWrapper {}`;
